data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java

```diff
@@ -338,9 +338,13 @@ public void run() {
         boolean retryingAfterException = false;
         while (!shutdownInProgress.get()) {
             try {
-                if (retryingAfterException || pauseConsumePredicate.pauseConsuming()) {
-                    LOG.debug("Pause consuming from Kafka topic.");
+                if (retryingAfterException) {
+                    LOG.debug("Pause consuming from Kafka topic due a previous exception.");
                     Thread.sleep(10000);
+                } else if (pauseConsumePredicate.pauseConsuming()) {
+                    LOG.debug("Pause and skip the next consume from Kafka topic due to an external condition: {}", pauseConsumePredicate);
+                    Thread.sleep(10000);
+                    continue;
```

Comment:
kafka consumer max.poll.interval will expire if we continue to pause for longer duration. This problem looks similar to https://github.com/opensearch-project/data-prepper/pull/4023 where bufferAccumator gets full and we explicitly pause kafka consumer so that max poll does not expire. Not sure if we are taking in the PR (kkondaka). May be we can combine these 2 conditions and have a common solution.

Reply:
@hshardeesi , This is a good point. I'll take another look at that angle. Thanks!

Reply:
@hshardeesi , I pushed a change that pauses the Kafka consumer. I ended up calling it in a different location because this pause happens before any `poll` calls whereas the other happens while writing to the buffer.

Reply:
Good catch @hshardeesi I misunderstood it as part of shutdown handling.

Reply:
functionally this looks correct except that in other case (https://github.com/opensearch-project/data-prepper/pull/4023) we don't pause consumer immediately but delay it until we are close to expiring kafka max.poll timeout. I feel this pause/resume logic should be abstracted out into separate class, similar to pausePredicate. abstraction can maintain state on when & what partitions to pause/resume etc and the core logic remains at one place.

Reply:
@hshardeesi , These are all good points. I agree with consolidating the logic to make it consistent. Do we need that in this PR, or can we follow this on?